exchanges/okx/ws_requests.go

```diff
@@ -0,0 +1,364 @@
+
+	var resp []*SpreadOrderResponse
+	if err := ok.SendAuthenticatedWebsocketRequest(ctx, cancelSpreadOrderEPL, id, "sprd-cancel-order", []map[string]string{arg}, &resp); err != nil {
+		return nil, mergeErrorDetails(err, resp)
+	}
+
+	return extractSingleItem(resp)
+}
+
+// WSCancelAllSpreadOrders cancels all spread orders and return success message through the websocket channel.
+func (ok *Okx) WSCancelAllSpreadOrders(ctx context.Context, spreadID string) error {
+	arg := make(map[string]string, 1)
+	if spreadID != "" {
+		arg["sprdId"] = spreadID
+	}
+
+	id := strconv.FormatInt(ok.Websocket.AuthConn.GenerateMessageID(false), 10)
+
+	var resp []*ResponseResult
+	if err := ok.SendAuthenticatedWebsocketRequest(ctx, cancelAllSpreadOrderEPL, id, "sprd-mass-cancel", []map[string]string{arg}, &resp); err != nil {
+		return mergeErrorDetails(err, resp)
+	}
+
+	single, err := extractSingleItem(resp)
+	if err != nil {
+		return err
+	}
+
+	if !single.Result {
+		return errCancelAllSpreadOrdersFailed
+	}
+
+	return nil
+}
+
+// SendAuthenticatedWebsocketRequest sends a websocket request to the server
+func (ok *Okx) SendAuthenticatedWebsocketRequest(ctx context.Context, epl request.EndpointLimit, id, operation string, payload, result any) error {
+	if operation == "" || payload == nil {
+		return errInvalidWebsocketRequest
+	}
+
+	outbound := &struct {
+		ID        string `json:"id"`
+		Operation string `json:"op"`
+		Arguments any    `json:"args"`
+		// TODO: Add ExpTime to the struct, the struct should look like this:
+		// ExpTime  string `json:"expTime,omitempty"` so a deadline can be set
+		// Request effective deadline. Unix timestamp format in milliseconds, e.g. 1597026383085
+	}{
+		ID:        id,
+		Operation: operation,
+		Arguments: payload,
+	}
+
+	incoming, err := ok.Websocket.AuthConn.SendMessageReturnResponse(ctx, epl, id, outbound)
+	if err != nil {
+		return err
+	}
+
+	intermediary := struct {
+		ID        string `json:"id"`
+		Operation string `json:"op"`
+		Code      int64  `json:"code,string"`
+		Message   string `json:"msg"`
+		Data      any    `json:"data"`
+		InTime    string `json:"inTime"`
+		OutTime   string `json:"outTime"`
+	}{
+		Data: result,
+	}
+
+	if err := json.Unmarshal(incoming, &intermediary); err != nil {
+		return err
+	}
+
+	switch intermediary.Code {
+	case 0:
+		return nil
+	case 1:
+		return errOperationFailed
+	case 2:
+		return errPartialSuccess
+	default:
+		return getStatusError(intermediary.Code, intermediary.Message)
+	}
+}
+
+func mergeErrorDetails[T interface{ Error() error }](err error, resp []T) error {
+	var errs error
+	for i := range resp {
+		if err := resp[i].Error(); err != nil {
+			errs = common.AppendError(errs, fmt.Errorf("index: `%d` error: `%w`", i+1, err))
+		}
+	}
+	if errs == nil {
+		return err
+	}
+	if err == nil {
+		return errs
+	}
+	return fmt.Errorf("%w: %w", err, errs)
+}
+
+func extractSingleItem[T any](resp []*T) (*T, error) {
```

Comment:
🐒 Sorry I missed this before.
* Slightly strange naming with `extract` in this. `get` is better. I'm wary that docs recently recommended not having `get`, though
* `Element` over `Item` in Go docs too.
* sometimes we see `must` prefix in naming like this
* 🥃 In this context I'm fond of the adjective `Singleton`
So overall this is what feels most idiomatic in naming:
```suggestion
func singleElement[T any](resp []*T) (*T, error) {
```